### ISSUE TITLE
We need better logging

### ISSUE THREAD

**Issue by athas** [diku-dk/futhark#12]
Well, we need logging at all.  Currently it's really hard to debug the compiler, because you have to insert surgical calls to 'Debug.Trace' where you think things are going wrong.  It would be nice if the compiler itself kept a running log of most of the things it rewrites (and maybe even why).  The following problems arise:
- Due to renaming, it may be hard to follow what is going on.  I've been minimising the amount of renaming going on, so maybe this isn't a big problem.  It generally is not when I trace.
- The code for performing the logging may be large and obscure the actual optimisations.  This can be reduced by design improvements - for example, if we require that each simplification rule is associated with a name (instead of being a plain Haskell function), we can in one place put a log message that says "rule R turned X into Y".
- When turned off, the deactivated logging code must be completely free of cost.  Lazy evaluation and monadic design will help us with this.

I do not want to create an `#ifdef` nightmare.  The degree of logging must be a command-line parameter - maybe even supporting some simple filters.  Also, the logging code must be entirely pure - we currently do not have IO anywhere but a single module, and I would like to keep it that way.

**Comment by Athas** [comment]
I recently added a basic logging facility to the pipeline infrastructure.

**Comment by athas** [comment]
The logging is OK now.
